### PR TITLE
Discord interactions

### PR DESCRIPTION
**Pull request by Acnologla**
# Description

This pull request adds discord interactions gateway interface.

Example: 

```go
func interaction(session disgord.Session, evt *disgord.InteractionCreate) {
	if evt.Type == disgord.InteractionMessageComponent {
		if evt.Data.CustomID == "click_one" {
			txt := fmt.Sprintf("User %s clicked in the button %s", evt.Member.User.Username, evt.Message.Components[0].Components[0].Label)
			err := disgord.SendInteractionResponse(evt, &disgord.InteractionResponse{
				Type: disgord.ChannelMessageWithSource,
				Data: &disgord.InteractionApplicationCommandCallbackData{
					Content: txt,
				},
			})
			fmt.Println(err)
		}
	}
}
func main() {
	client := disgord.New(disgord.Config{
		BotToken:     "token",
	})
	defer client.Gateway().StayConnectedUntilInterrupted()
	client.Gateway().InteractionCreate(interaction)
}
```

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Benchmarks
If this PR requires benchmarks (say it is an very dependent component or takes a lot of resources/use, use pprof if you need to) then the benchmarks are provided before and after such that we can make logical decisions.
Note that if you add a benchmark and find your solution to run slower, the code might still be valuable so your results are welcomed anyways!
If no benchmarks are needed, feel free to delete til paragraph.

# Checklist:

- [X] I ran `go generate`
- [ ] I have performed a self-review of my own code (remember to run `go fmt ./...`)
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] Added benchmarks if this is a performant required component (potential bottlenecks)

### CODE REVIEW

File: interaction.go [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+	Flags           int              `json:"flags"`
+	AllowedMentions *AllowedMentions `json:"allowed_mentions"`
+}
+
+type InteractionResponse struct {
+	Type InteractionCallbackType                    `json:"type"`
+	Data *InteractionApplicationCommandCallbackData `json:"data"`
+}
+
+func SendInteractionResponse(interaction *InteractionCreate, data *InteractionResponse) error {
```
Comment (andersfylling) [inline interaction.go:88]: This needs to be a part of the client:
 1. rate limiting, if there is any
 2. re-use the http instance (proxy support + other things)

The way it's done in disgord now is a bit, over-engineered, and lately ive been trying to go away from it. However, see if you can create a `httd.Request` object and call `Client.req.Do(context.Context, *httd.Request)`. After the `Do` call, you don't need to close the body of the request and you can read directly from the returned body instead. Also no need to worry about error handling as this method does that for you.

**Comment by andersfylling** [comment]
Looks good! I have one request that needs to adjusted, but besides that it's all clean.

**Comment by andersfylling** [comment]
While you might see methods looking like this, some can easily be converted.
image

REFACTORED - I guess something like this is how it would look for your case
image

File: interaction.go (hunk 1 above)
Comment (Acnologla) [reply interaction.go]: Done

**Pull request merged**